OpenDAL S3 parameter support

txase opened the pull request:
This change allows s3:// data paths to include additional parameters as querystrings. The default parameters remain optimized for the official AWS S3 service, but any OpenDAL S3 parameters can be set. For example, to override the default `true` value for `enable_virtual_host_style` you can set the data config to:

s3://my-bucket?enable_virtual_host_style=false

Boolean parameters can be set to `true` by simply adding the parameter:

s3://my-bucket?enable_versioning

Fixes #6112

ProbstDJakob commented:
When this has been merged, it would be great if the official container images would be build with S3 support enabled.

kaiyou commented:
Thank you so much for this much expected feature.

I have tested this for a couple days and deployed it to production today. It works really well and also enables compatibility with Garage, yet another S3 cluster implementation.

I have feedback about the admin experience though: `ICON_CACHE_FOLDER`, `SENDS_FOLDER` and `ATTACHMENTS_FOLDER`, when not explicitly provided, are automatically derived from the more general `DATA_FOLDER` in `config.rs`. One might either set all of these explicitly, e.g. to store data in separate buckets, or simply set `DATA_FOLDER` to an S3 URL and let OpenDAL deal with derived subpaths.

However, when `DATA_FOLDER`  contains a query string, it cannot be concaternated with `/attachments`, `/sends` or `/icons_cache` anymore. For instance : 

```
s3://mybucket?endpoint=my-service.tld&region=myregion
```

... leads to the following attachment folder URL :

```
s3://mybucket?endpoint=my-service.tld&region=myregion/attachments
```

The subpath is now part of the region in the query string, which just fails (due to `/` being forbidden in region names and sigv4 scope construction). I suspect other situations might lead to more subtle failure modes, where requests are handled properly but data is directed to the bucket root instead of the intended subpath.

It would be really nice if the subpath was properly inserted before the query string. This either requires some more logic in `config.rs`, parsing and manipulating the OpenDAL URL, or a separate configuration variable for the query string. Otherwise, a note in `.env.template` would help avoid some failure modes.

txase commented:
> @txase i think a new release is coming soon, and a few PR's will probably be merged before this release. So, probably it's best after that release i think?

@BlackDex I saw some releases go by this past week. I'll be busy next week, but maybe can get to this the week of April 27th. Would that work well for you to slot in reviews to get this rebased, tested, and merged?

txase committed:
storage: add OpenDAL S3 URI options

OpenDAL S3 storage accepts bucket and root path data today, but
serverless deployments also need URI query parameters to describe provider
behavior in one DATA_FOLDER value.

Update OpenDAL to 0.56.0 and build S3 operators with
S3Config::from_uri(). Keep Vaultwarden's AWS SDK credential chain by
installing a reqsign provider when the URI does not explicitly request
OpenDAL-native credential handling.

Move path handling and operator construction into storage.rs so S3-specific
parsing, credential setup, and URI path manipulation stay out of
configuration handling. Local filesystem behavior is unchanged, and S3
child paths are derived before query strings.

txase commented:
I spent time rebasing and figuring out how to address @kaiyou's issue and incorporate the from_uri() config method that OpenDAL now provides. The result is this new commit. It is a bit larger due to the following:

- OpenDAL is updated to the latest version
- The latest version of OpenDAL has a different manner of setting up custom AWS request signing (we use the official AWS SDK for signing config instead of OpenDAL's implementation because it tends to lag behind AWS SDK functionality like newer SSO auth options)
  - That said, this new version is nicer in that it doesn't require adding the `anyhow` crate anymore and uses rustls for TLS
- It's a non-trivial change to solve @kaiyou's issue where params and path prefixes are mixed
  - To solve for this, I created a new `storage` module with an `s3`-specific submodule to abstract away the details
  
  This is now ready for review again!

txase committed:
deps: upgrade the reqwest stack to 0.13

The reqwest 0.13 rustls feature selects the aws-lc provider. Use
rustls-no-provider instead, add rustls 0.23 with the ring provider, and
install that provider at process startup. This keeps Vaultwarden on the
existing ring crypto provider while giving reqwest, OpenDAL and lettre a
process-wide rustls provider.

Disable openidconnect default features and provide a small
AsyncHttpClient wrapper around Vaultwarden's shared reqwest client
builder. This preserves custom DNS, request blocking, timeouts and the
no-redirect OIDC behavior without openidconnect enabling its own reqwest
stack.

Upgrade yubico_ng to 0.15.0 and OpenDAL to 0.56.0. OpenDAL 0.56 also
moves S3 signing to reqsign 3, so switch the optional S3 dependencies
from reqsign/anyhow to reqsign-core and reqsign-aws-v4 and adapt the AWS
SDK credential bridge to the new ProvideCredential API.

Adjust the local OpenDAL call sites for the 0.56 API: use the FS_SCHEME
constant for filesystem checks and replace deprecated remove_all() with
delete_with(...).recursive(true) for Send file cleanup.

BlackDex reviewed (approved):
Nice, seems to work like a charm, and a lot of dependencies are gone/removed!

dani-garcia reviewed (approved):
Looks good to me too, thanks!
